### PR TITLE
Scrabble Assignment 

### PR DESCRIPTION
**Pull request by amanzan1**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | The advantages are that you can view your changes on git through git log and push changes to your partner's computer. Git is able to some extent merge both of you and your partner's changes.
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | I think because it does not have a state to remember.
Describe an example of a test you wrote for a _nominal case_. | One test we wrote was checking to see if tiles drawn were being removed from the tile_bag, so that actual number of tiles equaled the expected amount of tiles, which was a default number.
Describe an example of a test you wrote for an _edge case_. | An example would be writing  a test that would return nil if no words were passed. We created a word collection variable that was passed into the highest.score_from method and requiring that output to be nil.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | No.
What was one method you and your pair used to debug code? | Running the rake command and using the puts method to print out what the program is doing.
Is there a specific piece of code you'd like feedback on? | I can't think of anything  at the moment.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We thought the process went smoothly.

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,97 @@
...
+      # returns nil if word is empty
+      if letters_array.empty?
+        return nil
+      end
+
+      # original word score
+      letters_array.each do |letter|
+        if ("a".."z").cover?(letter)
+          case
+          when letter == "d" || letter == "g"
```
Comment (CheezItMan) [inline lib/scoring.rb:26]: This structure isn't the most efficient way to do it.  Can you see a way to do this more compactly with a hash?

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,97 @@
...
+      # word-length scoring bonus
+      if letters_array.length == 7
+        total += 50
+      end
+
+      return total
     end
 
     def self.highest_score_from(array_of_words)
+      array_of_words = array_of_words
```
Comment (CheezItMan) [inline lib/scoring.rb:55]: This line doesn't actually do anything.

File: lib/scoring.rb [hunk 3]
```diff
@@ -1,9 +1,97 @@
...
+        return array_of_words[0]
+      end
+
+      array_of_words.each do |word|
+        value = Scrabble::Scoring.score(word)
+        if value > highest_score
+          highest_score = value
+          highest_word = word
+
+        elsif (value == highest_score) && (word.length == 7)
```
Comment (CheezItMan) [inline lib/scoring.rb:73]: You also need to check that `highest_word` isn't 7-letters long.  You're missing a case.

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -42,21 +42,49 @@
...
+      Scrabble::Scoring.highest_score_from(word_collection2).must_equal "mouse"
     end
 
     it 'returns the highest word if there are two words' do
+      word_collection = ["dog","quiz","cat"]
+      Scrabble::Scoring.highest_score_from(word_collection).must_equal "quiz"
     end
 
     it 'if tied, prefer a word with 7 letters' do
+      #Act
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:66]: What about having 2 tied 7-letter words?

File: specs/player_spec.rb [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+describe 'Player' do
+
+  describe 'new' do
+    it 'returns the value of the @name instance variable' do
+      jim = Scrabble::Player.new("Jim")
+      jim.name.must_equal "Jim"
+    end
+  end
+
+  describe 'plays' do
```
Comment (CheezItMan) [inline specs/player_spec.rb:20]: You should also try to play words after the player has already won.

File: lib/tilebag.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+      @starting_tiles = {
+        A: 9, B: 2, C: 2, D: 4, E: 12, F: 2,
+        G: 3, H: 2, I: 9, J: 1, K: 1, L: 4,
+        M: 2, N: 6, O: 8, P: 2, Q: 1, R: 6, S: 4,
+        T: 6, U: 4, V: 2, W: 2, X: 1, Y: 2, Z: 1
+      }
+
+      @remaining_tiles = []
+      @starting_tiles.each do |letter,value|
```
Comment (CheezItMan) [inline lib/tilebag.rb:20]: Really good that you do this.   An array is a great structure to use.

File: lib/tilebag.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+require 'pry'
+require "awesome_print"
+
+module Scrabble
+  class TileBag
+    attr_accessor :starting_tiles, :hand_of_tiles, :remaining_tiles
+
+    def initialize
+
+      @starting_tiles = {
```
Comment (CheezItMan) [inline lib/tilebag.rb:12]: `starting_tiles` would make better sense as a constant.

File: lib/tilebag.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+    def tiles_remaining
+       return @remaining_tiles.length
+    end
+
+  end
+end
+
+
+first = Scrabble::TileBag.new
```
Comment (CheezItMan) [inline lib/tilebag.rb:42]: Please take out these pieces of test code out when you submit a project.

File: lib/tilebag.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        value.times do
+          @remaining_tiles.push(letter.to_s)
+        end
+      end
+
+      @remaining_tiles = @remaining_tiles.shuffle
+
+    end
+
+    def draw_tiles(number)
```
Comment (CheezItMan) [inline lib/tilebag.rb:30]: What about when the user tries to draw more tiles than remain in the tile bag.

File: specs/tilebag_spec.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    end
+
+    it 'removes used tiles from the default set' do
+
+      grab_bag = Scrabble::TileBag.new
+
+      original_number_of_tiles = grab_bag.tiles_remaining
+      grab_bag.draw_tiles(6)
+
+      expected_number_of_tiles = original_number_of_tiles - 6
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:38]: good test

File: specs/tilebag_spec.rb [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      actual_number_of_tiles = grab_bag.tiles_remaining
+
+      actual_number_of_tiles.must_equal expected_number_of_tiles
+
+    end
+
+  end
+
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:46]: You should also test drawing more tiles than remain in the bag.

File: specs/tilebag_spec.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  xdescribe 'tiles_remaining' do
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:47]: What is this test skipped?

File: specs/tilebag_spec.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    it 'returns the number of tiles remaining in the bag' do
+      grab_bag = Scrabble::TileBag.new
+
+      number_of_tiles = grab_bag.tiles_remaining
+
+      grab_bag.remaining_tiles.length.must_equal number_of_tiles
```
Comment (CheezItMan) [inline specs/tilebag_spec.rb:53]: better to hard-code the number of tiles.

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Check, You can't actually push changes to your partner's computer.  Instead you push changes up to github and they can pull the changes down to their computer.  
Both Teammates contributed to the codebase | Both teammates made commits!
Regular Commits with meaningful commit messages |  Good number of commits, but focus more on what code you wrote in the commits and less on the waves completed.  That would make it easier to rollback a commit if you need to later.  
**Scoring class** |
`score` method |  Check
Uses appropriate data structure to store the letter score |  Check, although the structure isn't the most efficient way to do it. See my inline notes.
Appropriately handles edge cases |  Check
`highest_score_from` method |  Check
Appropriately handles edge cases |  Small bug here, see my inline notes
Test implementations for `highest_score_from`|  Missing a test for tied words and both are 7-letters long.  Your code returns the last word in case of a tie, if it's 7-letters long without considering there might be a high-scoring word that's seven letters long earlier.  
**Player class** |
Uses an array to keep track of words played | Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` | Check
Returns `true` or `false` from `won?` method |  Check
Tests edge cases on the `play(word)` | Pretty good, see my notes.  
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag | **Very** good choice in using an Array
`draw_tiles` method uses the parameter to draw a set of random tiles |  Check, but see my note on drawing too many tiles.
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  You didn't test drawing more tiles than are in the bag.
Summary | There is some good code here and some good testing.  There are some serious bugs, notably in the `highest_score_from`, so look at that and look at how your tests missed that situation.  You need to look at testing like a detective.  Try to think about what **could** go wrong with your code.  Read my notes and let me know any questions you have.
